third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -2730,19 +2826,396 @@ impl Frame {
+        use LoadedFunctionOwner::*;
+        let instantiation = match self.function.owner() {
+            Module(module) => module.function_instantiation_at(idx.0),
+            Script(script) => script.function_instantiation_at(idx.0),
+        };
+
+        let ty_args = self.function.ty_args();
+        if let Some(gas_meter) = gas_meter {
+            for ty in instantiation {
+                gas_meter
+                    .charge_create_ty(NumTypeNodes::new(ty.num_nodes_in_subst(ty_args)? as u64))?;
+            }
+        }
+
+        let instantiation = instantiation
+            .iter()
+            .map(|ty| self.ty_builder.create_ty_with_subst(ty, ty_args))
+            .collect::<PartialVMResult<Vec<_>>>()?;
+        Ok(instantiation)
+    }
+}
+
+macro_rules! build_loaded_function {
+    ($function_name:ident, $idx_ty:ty, $get_function_handle:ident) => {
+        pub(crate) fn $function_name(
+            &self,
+            module_storage: &impl ModuleStorage,
+            idx: $idx_ty,
+            verified_ty_args: Vec<Type>,
+        ) -> PartialVMResult<LoadedFunction> {
+            match self.function.owner() {
+                LoadedFunctionOwner::Module(module) => {
+                    let handle = module.$get_function_handle(idx.0);
+                    match handle {
+                        FunctionHandle::Local(function) => {
+                            (Ok(LoadedFunction {
+                                owner: LoadedFunctionOwner::Module(module.clone()),
+                                ty_args: verified_ty_args,
+                                function: function.clone(),
+                            }))
+                        },
+                        FunctionHandle::Remote { module, name } => self
+                            .build_loaded_function_from_name_and_ty_args(
+                                module_storage,
+                                module,
+                                name,
+                                verified_ty_args,
+                            ),
+                    }
+                },
+                LoadedFunctionOwner::Script(script) => {
+                    let handle = script.$get_function_handle(idx.0);
+                    match handle {
+                        FunctionHandle::Local(_) => Err(PartialVMError::new(
+                            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                        )
+                        .with_message("Scripts never have local functions".to_string())),
+                        FunctionHandle::Remote { module, name } => self
+                            .build_loaded_function_from_name_and_ty_args(
+                                module_storage,
+                                module,
+                                name,
+                                verified_ty_args,
+                            ),
+                    }
+                },
+            }
+        }
+    };
+}
+
+impl Frame {
```

Comment:
Above was also Impl of Frame right?
Can we take the opportunity to move frame out of this file? That would be great.. might also be a great opportunity to add a few tests to it, just to get them started.

Reply:
Moved to a new file `frame.rs` (will be fun to re-stack my PR above💀)